### PR TITLE
Support Implicit Namespace Packages (PEP 420)

### LINKED ISSUE #7959: PEP420 `cyclic-import` false positive
### Bug description

https://github.com/alexey-pelykh/reproduce-pylint-pep420-cyclic-import-false-positive

`pylint` thinks that `logging` is the package name while due to implicit namespace, it's `some.package.logging`.

### Configuration

```ini
[MASTER]
init-hook=
    from os.path import dirname, join
    from glob import glob
    from pylint.config import find_pylintrc
    from site import addsitedir
    site_packages_paths = glob(join(dirname(find_pylintrc()), '**', '*venv', '**', 'site-packages'), recursive=True)
    for site_packages_path in site_packages_paths: addsitedir(site_packages_path)
```

### Command used

```shell
./venv/bin/pylint --recursive=y src
./venv/bin/pylint $(find src -name '*.py')
```

### Pylint output

```shell
************* Module logging.wrapper.__init__
src/some/package/logging/wrapper/__init__.py:1:0: R0401: Cyclic import (logging -> logging.wrapper) (cyclic-import)
```

### Expected behavior

No cyclic import detected

### Pylint version

```shell
pylint 2.15.9
astroid 2.12.13
Python 3.11.0 (main, Oct 26 2022, 19:06:18) [Clang 14.0.0 (clang-1400.0.29.202)]
```

### OS / Environment

_No response_

### Additional dependencies

_No response_

### LINKED ISSUE #8154: Support Implicit Namespace Packages (PEP 420)
### Current problem

(#7959)

### Desired solution

Support Implicit Namespace Packages (PEP 420)

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by alexey-pelykh**
Fixes #7959 by introducing the concept of source roots.

## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :bug: Bug fix          |
| ✓   | :sparkles: New feature |
| ✓   | :hammer: Refactoring   |

## Description

Closes #8154

### CODE REVIEW

**Comment by alexey-pelykh** [comment]
@Pierre-Sassoulas so it is done and it would be great to get the feedback so I can address it promptly :)

**Review by Pierre-Sassoulas — COMMENTED** [review]
I'm wondering if we could use the packaging information to discover the source root. 

It's possible that making the user give the source root is the way to fix this once and for all. I did not review everything yet. I hope some other maintainer will have an opinion about this. 

Changes to the namespace and file discovery are very frightening because there's not a lot of way to test it properly on the scale it needs to be tested to be sure it's not going to explode everywhere when we release it. So I'm very conservative when modifying this kind of code and I think we need extensive tests, maybe more (small) primers with implicit namespace package and various use cases.

File: tests/lint/unittest_lint.py
Comment (jacobtylerwalls) [inline tests/lint/unittest_lint.py:1278]: Can you add more context around why these tests needed to change? If there's a breaking change for users, we need to understand it and document it.

Comment (jacobtylerwalls) [reply tests/lint/unittest_lint.py:1278]: OH! Is is just that `fix_import_path` is being deprecated? If so, I get it now.

Comment (alexey-pelykh) [reply tests/lint/unittest_lint.py:1278]: Yep, that's right :)

**Comment by alexey-pelykh** [comment]
@Pierre-Sassoulas thanks for your feedback and please find my comments below:

> I'm wondering if we could use the packaging information to discover the source root.

A good point, absolutely. Yet that should serve as one a source of source roots information if the pylint configuration misses one and is an improvement to the current implementation. Also, such auto-magic would be a breaking change since projects would need an upgrade, as minor upgrade would bring behavior change. Probably a change for 3.0 no less.

> Changes to the namespace and file discovery are very frightening because there's not a lot of way to test it properly on the scale it needs to be tested to be sure it's not going to explode everywhere when we release it. So I'm very conservative when modifying this kind of code and I think we need extensive tests, maybe more (small) primers with implicit namespace package and various use cases.

The current behavior remains unchanged so there shouldn't be a disaster once released and the test suite was expanded as well the new optional behavior.

**Comment by DudeNr33** [comment]
Thank you @alexey-pelykh for this work and especially for also thinking about `pyreverse` 🎉

I agree with Pierre that adding some projects that use implicit namespace packages would be a good idea.

Besides that I was only wondering why it is necessary to create several new functions and deprecate the existing ones, when the now deprecated functions afterwards just delegate to the new functions.

**Comment by alexey-pelykh** [comment]
@DudeNr33 thanks for your feedback and please find my comments below:

> I agree with Pierre that adding some projects that use implicit namespace packages would be a good idea.

Now I'm not sure I followed the idea correctly: adding projects as a test fixtures? adding where and how?
I'd be happy to do that, yet I'm not quite sure _what exactly_ should I do :)

> Besides that I was only wondering why it is necessary to create several new functions and deprecate the existing ones, when the now deprecated functions afterwards just delegate to the new functions.

Both `fix_import_path` and `_patch_sys_path` are exposed as a public API, therefore possibly are used by someone, so I couldn't add a parameter w/o default value or rename. Adding a parameter with default value while would keep the API backwards compatible, won't alert the possible users in any way that it's better to provide source root(s) explicitly to avoid pylint doing guesswork. The renaming while optional really felt right since it's not "fixing" the path, it's extending or augmenting it.

That said, the most robust solution would be to warn users that used API causes implicit guesswork via deprecation and provide an explicit way to force pylint do the guesswork via new method signatures. As a nice bonus, some naming adjustments could be be made. And in 3.0 release those deprecated methods would be removed.

**Comment by DudeNr33** [comment]
Thank you for the explanation of the deprecation warnings. I agree that this is then the cleaner approach!
 
> Now I'm not sure I followed the idea correctly: adding projects as a test fixtures? adding where and how? I'd be happy to do that, yet I'm not quite sure _what exactly_ should I do :)

[Primer tests](https://pylint.pycqa.org/en/latest/development_guide/contributor_guide/tests/launching_test.html#primer-tests) run `pylint` against a selection of different repositories to check

* if `pylint` crashes
* if the changes of a pull request lead to a change in the messages generated, to help finding false positives for example

Unfortunately I have no further experience in dealing with the primer test suite, so the other maintainers can help you more with that.

**Comment by alexey-pelykh** [comment]
Thanks for explanation on primer tests, makes sense. Yet no open-source project that use PEP420 with 2+ depth comes to my mind really...

**Comment by DudeNr33** [comment]
> Thanks for explanation on primer tests, makes sense. Yet no open-source project that use PEP420 with 2+ depth comes to my mind really... Yet thus far nothing have changed for projects currently listed for the primer tests, right?

[poetry-core](https://github.com/python-poetry/poetry-core) uses implicit namespace packages, but only one level deep. Nevertheless it could be a good candidate. 

Yes, for the existing primer projects all seems to be fine.

**Comment by alexey-pelykh** [comment]
> [poetry-core](https://github.com/python-poetry/poetry-core) uses implicit namespace packages, but only one level deep.

It is an ideal one as is appears, exactly what I was looking for. So I'll add it :)

**Comment by alexey-pelykh** [comment]
@Pierre-Sassoulas any insights what was this about? https://github.com/PyCQA/pylint/blob/main/pylint/testutils/_primer/primer_run_command.py#L78

Also, it feels like https://github.com/PyCQA/pylint/pull/8193 needs to go into main before this one to properly capture unchanged poetry-core priming?

**Comment by Pierre-Sassoulas** [comment]
> @Pierre-Sassoulas any insights what was this about? https://github.com/PyCQA/pylint/blob/main/pylint/testutils/_primer/primer_run_command.py#L78

duplicate code is really slow, intrinsically solving a hard problem that scale pretty badly on big code base and the code almost never change so we removed it. For cyclic import I don't remember but maybe it's also slowness, maybe also that the result are not the same each run (?)

**Comment by alexey-pelykh** [comment]
@Pierre-Sassoulas please rerun https://github.com/PyCQA/pylint/actions/runs/4102946645/jobs/7076537678 otherwise nothing will happen

**Comment by Pierre-Sassoulas** [comment]
Trying to fix this in #8200

**Comment by alexey-pelykh** [comment]
Well, I guess all is in check now?

File: doc/user_guide/usage/run.rst
Comment (DudeNr33) [inline doc/user_guide/usage/run.rst:53]:
```suggestion
If the analyzed sources use implicit namespace packages (PEP 420), the source root(s) should
be specified using the ``--source-roots`` option. Otherwise, the package names are
detected incorrectly, since implicit namespace packages don't contain an ``__init__.py``.
```

File: doc/whatsnew/fragments/8154.feature
Comment (DudeNr33) [inline doc/whatsnew/fragments/8154.feature:1]:
```suggestion
Support implicit namespace packages (PEP 420).
```

**Review by DudeNr33 — APPROVED** [review]
I only have a few wording suggestions, but they are no blocker imo.
However I think at least one other maintainer should additionally approve before merging.

File: pylint/lint/utils.py
Comment (Pierre-Sassoulas) [inline pylint/lint/utils.py:80]: We could announce a removal in 3.0, maybe 4.0 seeing that 3.0 is nearing.

**Review by Pierre-Sassoulas — APPROVED** [review]
Thank you for all the work, clearly a lot of care went into this. I don't see anything outrageous and there are tests that I think are reasonable. Let's wait a little for more reviews. Then we could release a 2.17 beta version to make sure there's no obvious issues before really releasing in the wild :)

**Review by DudeNr33 — APPROVED** [review]
Looks good to me. 😊

**Pull request merged**
